packages/modal/src/Modal.tsx

```diff
@@ -0,0 +1,259 @@
+import React, { useCallback, SetStateAction, useRef } from 'react';
+import PropTypes from 'prop-types';
+import Portal from '@leafygreen-ui/portal';
+import { Global } from '@emotion/core';
+import Icon, { Size } from '@leafygreen-ui/icon';
+import { useEventListener } from '@leafygreen-ui/hooks';
+import { uiColors } from '@leafygreen-ui/palette';
+import { css, cx } from '@leafygreen-ui/emotion';
+
+export const ModalSize = {
+  XXSmall: 'xxsmall',
+  XSmall: 'xsmall',
+  Small: 'small',
+  Default: 'default',
+  Large: 'large',
+  XLarge: 'xlarge',
+} as const;
+
+export type ModalSize = typeof ModalSize[keyof typeof ModalSize];
+
+const defaultSpacing = 18;
+
+const scrollContainer = css`
+  animation: fade-in 500ms cubic-bezier(0.165, 0.84, 0.44, 1);
+  -webkit-animation: fade-in 500ms cubic-bezier(0.165, 0.84, 0.44, 1);
+  background-color: ${uiColors.black};
+  opacity: 0.6;
+  overflow-y: auto;
+  position: fixed;
+  top: 0;
+  left: 0;
+  right: 0;
+  bottom: 0;
+  display: flex;
+  justify-content: center;
+  align-items: flex-start;
+  pointer-events: initial;
+`;
+
+const modalContentStyle = css`
+  margin: ${defaultSpacing}px auto;
+  max-width: 1270px;
+  min-width: 944px;
+  padding: 30px;
+  color: ${uiColors.gray.dark3};
+  background-color: ${uiColors.white};
+  border: 1px solid ${uiColors.gray.light3};
+  border-radius: 3px;
+  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.3);
+  pointer-events: all;
+  height: 200vh;
+`;
+
+const modalBodyStyle = css`
+  padding: ${defaultSpacing}px 0;
+`;
+
+const modalSizes: { readonly [K in ModalSize]: string } = {
+  xxsmall: css`
+    min-width: inherit;
+    max-width: inherit;
+    width: 366px;
+  `,
+
+  xsmall: css`
+    min-width: inherit;
+    max-width: inherit;
+    width: 418px;
+  `,
+
+  small: css`
+    min-width: inherit;
+    max-width: inherit;
+    width: 562px;
+  `,
+
+  default: css`
+    min-width: inherit;
+    max-width: inherit;
+    width: 700px;
+  `,
+
+  large: css`
+    min-width: inherit;
+    max-width: inherit;
+    width: 800px;
+  `,
+
+  xlarge: css`
+    min-width: inherit;
+    max-width: inherit;
+    width: 1270px;
+  `,
+};
+
+const closeButton = css`
+  color: ${uiColors.gray.dark1};
+  float: right;
+  cursor: pointer;
+`;
+
+const escapeKey = 27;
+
+interface ModalProps {
+  /**
+   * Content that will appear inside of the Modal component.
+   */
+  children: React.ReactNode;
+
+  /**
+   * Determines the active state of the modal
+   *
+   * default: `false`
+   */
+  active: boolean;
+
+  /**
+   * Specifies the size of the Modal.
+   *
+   * default: `default`
+   */
+  size?: ModalSize;
+
+  /**
+   * Callback to change the active state of the Modal.
+   *
+   */
+  setActive?: (
+    active: boolean,
+  ) => void | React.Dispatch<SetStateAction<boolean>>;
+
+  /**
+   * Callback to determine whether or not Modal should close when user tries to close it.
+   *
+   */
+  modalShouldClose?: () => boolean;
```

Comment:
Maybe in a future release if people desire, we should include an object to the function as a reason for why we're asking if the Modal should close. i.e. `{reason: 'escapeKey Pressed' }`, such that a user can make smarter decisions in this function.

Reply:
Also a rename to `shouldClose`? or `canClose`?

Reply:
I can rename to shouldClose, I know we were attempting to imitate the componentShouldUpdate name, but shouldClose works for me as well if you're good with it @DesignerDave

Reply:
Yeah, seems fine. I also like the "exit code" idea